File: src/main/java/xyz/phanta/cbtweaker/integration/jei/ingredient/JeiForgeEnergyIngredient.java
package xyz.phanta.cbtweaker.integration.jei.ingredient;

import net.minecraft.client.util.ITooltipFlag;
import xyz.phanta.cbtweaker.gui.CbtTextureResources;

import java.util.Collections;
import java.util.List;

public class JeiForgeEnergyIngredient implements JeiIngredient<Integer> {

    private final int amount;
    private final String unitSymbol;
    private final Role role;

    public JeiForgeEnergyIngredient(int amount, String unitSymbol, Role role) {
        this.amount = amount;
        this.unitSymbol = unitSymbol;
        this.role = role;
    }

    public int getAmount() {
        return amount;
    }

    @Override
    public List<Integer> getIngredients() {
        return Collections.singletonList(amount);
    }

    @Override
    public Role getRole() {
        return role;
    }

    @Override
    public void renderIcon(int x, int y, Integer ingredient) {
        CbtTextureResources.ICON_ENERGY.draw(x, y);
    }

    @Override
    public void getTooltip(Integer ingredient, List<String> tooltip, ITooltipFlag tooltipFlags) {
        tooltip.add(String.format("%,d %s", ingredient, unitSymbol));
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/compat/jei/ingredient/impl/JeiForgeEnergyIngredient.kt
package st.evening.mc.cbtweaker.compat.jei.ingredient.impl

import net.minecraft.client.util.ITooltipFlag
import net.minecraft.util.text.TextFormatting
import st.evening.mc.cbtweaker.compat.jei.ingredient.JeiIngredient
import st.evening.mc.cbtweaker.gui.CbtGuiResources
import st.evening.mc.prelude.api.gui.drawable.drawFullSize
import st.evening.mc.prelude.api.util.game.ClientSide
import st.evening.mc.prelude.api.util.text.toStringPercentage

class JeiForgeEnergyIngredient(
    val amount: Int,
    val unitName: String,
    override val role: JeiIngredient.Role,
    val chance: Float = 1F
) : JeiIngredient<Int> {
    override fun getIngredients(): List<Int> = listOf(amount)

    @ClientSide.Physical
    override fun drawIcon(x: Int, y: Int, ingredient: Int, partialTicks: Float) {
        CbtGuiResources.ICON_ENERGY.drawFullSize(partialTicks, x, y)
    }

    @ClientSide.Physical
    override fun getTooltip(ingredient: Int, tooltip: MutableList<String>, tooltipFlags: ITooltipFlag) {
        tooltip += "%,d %s".format(ingredient, unitName)
        if (chance < 1F) {
            tooltip += "${TextFormatting.GOLD}(${chance.toStringPercentage()})"
        }
    }
}
